src/workerd/server/channel-token.c++

```diff
@@ -0,0 +1,289 @@
+    IoChannelFactory::ChannelTokenUsage usage,
+    kj::StringPtr serviceName,
+    kj::Maybe<kj::StringPtr> entrypoint,
+    Frankenvalue& props) {
+  capnp::word scratch[128]{};
+  capnp::MallocMessageBuilder message(scratch);
+  auto builder = message.getRoot<ChannelToken>();
+
+  builder.setType(type);
+
+  // If requireAllowsTransfer() passed, then we are not dynamic so should have a service name.
+  builder.setName(serviceName);
+
+  KJ_IF_SOME(e, entrypoint) {
+    builder.setEntrypoint(e);
+  }
+
+  {
+    auto propsBuilder = builder.initProps();
+    props.toCapnp(propsBuilder);
+
+    auto capTable = props.getCapTable();
+    if (capTable.size() > 0) {
+      auto tableBuilder = propsBuilder.initCapTable().initAs<ChannelToken::FrankenvalueCapTable>();
+
+      auto caps = tableBuilder.initCaps(capTable.size());
+
+      for (auto i: kj::indices(capTable)) {
+        KJ_IF_SOME(subreq, kj::tryDowncast<IoChannelFactory::SubrequestChannel>(*capTable[i])) {
+          caps[i].setSubrequestChannel(subreq.getToken(usage));
+        } else KJ_IF_SOME(actorClass,
+            kj::tryDowncast<IoChannelFactory::ActorClassChannel>(*capTable[i])) {
+          caps[i].setActorClassChannel(actorClass.getToken(usage));
+        } else {
+          KJ_FAIL_REQUIRE("unknown type in props");
+        }
+      }
+    }
+  }
+
+  kj::VectorOutputStream out;
+  capnp::writePackedMessage(out, message);
+
+  auto plaintext = out.getArray();
+
+  switch (usage) {
+    case IoChannelFactory::ChannelTokenUsage::RPC: {
+      auto result = kj::heapArray<byte>(sizeof(TokenHeader) + plaintext.size() + AES_MAC_SIZE);
+      auto& header = *reinterpret_cast<TokenHeader*>(result.begin());
+
+      header.magic = ChannelToken::RPC_TOKEN_MAGIC;
+      getEntropy(header.iv);
+      kj::arrayPtr(header.keyId).copyFrom(keyId);
+
+      EVP_CIPHER_CTX* aesCtx = EVP_CIPHER_CTX_new();
+      KJ_ASSERT(aesCtx != nullptr);
+      KJ_DEFER(EVP_CIPHER_CTX_free(aesCtx));
+
+      KJ_ASSERT(EVP_EncryptInit(aesCtx, EVP_aes_256_gcm(), tokenKey, header.iv));
+
+      // Add header as AAD first.
+      {
+        int outSize = 0;
+        KJ_ASSERT(
+            EVP_EncryptUpdate(aesCtx, nullptr, &outSize, result.begin(), sizeof(TokenHeader)));
+        KJ_ASSERT(outSize == sizeof(TokenHeader));
+      }
+
+      // Encrypt the body.
+      {
+        int outSize = 0;
+        KJ_ASSERT(EVP_EncryptUpdate(aesCtx, result.begin() + sizeof(TokenHeader), &outSize,
+            plaintext.begin(), plaintext.size()));
+        KJ_ASSERT(outSize == plaintext.size());  // because AES-GCM is a stream cipher
+      }
+
+      int out = 0;
+      KJ_ASSERT(EVP_EncryptFinal_ex(aesCtx, nullptr, &out));
+      KJ_ASSERT(out == 0);  // No padding for stream ciphers like AES-GCM.
+
+      // Get the MAC.
+      KJ_ASSERT(EVP_CIPHER_CTX_ctrl(
+          aesCtx, EVP_CTRL_GCM_GET_TAG, AES_MAC_SIZE, result.end() - AES_MAC_SIZE));
+
+      return result;
```

Comment:
Just out of curiosity, what are the security requirements for these tokens? I assume there's not much needed here. Might be good to document briefly tho. Non-blocking.